### DISCUSSION TITLE
IState binding error Uno.Extensions.Reactive.Core.StateImpl`1

### DISCUSSION THREAD

**Discussion by limeniye** [unoplatform/uno#11183] [Q&A]
I'm tring to use IState binding
I try to do as written in the examples.
The code is generated for me, but for some reason it does not work.

First way:
```C#
    public partial class AuthorizationViewModel
    {
        public IState<string> Login => State<string>.Empty(this);
    }
```

```XML
   <TextBox Text="{Binding Login, Mode=TwoWay, UpdateSourceTrigger=PropertyChanged}"/>
```

Result is next
![зображення](https://user-images.githubusercontent.com/69314237/216569878-ffef71f9-2962-487f-8fd3-39ab8a2ba800.png)

Second way:
```C#
    public partial class AuthorizationViewModel
    {
        public IState<Credentials> Credentials => State<Credentials>.Empty(this);

        public ICommand SignIn => Command.Create(b => b.Given(Credentials).When(CanSignIn).Then(DoSignIn));
        
        private bool CanSignIn(Credentials credentials)
		=> credentials is { UserName.Length: > 0 } and { Password.Length: > 0 };

	private async ValueTask DoSignIn(Credentials credentials, CancellationToken ct)
		=> await _navigator.NavigateViewModelAsync<FuturesViewModel>(this);
    }
```

```XML
    <TextBox Text="{Binding Credentials.UserName, Mode=TwoWay, UpdateSourceTrigger=PropertyChanged}"/>
    <PasswordBox Password="{Binding Credentials.Password, Mode=TwoWay, UpdateSourceTrigger=PropertyChanged}"/>
    <Button Command="{Binding SignIn}"/>
```

![зображення](https://user-images.githubusercontent.com/69314237/216572941-c40337cc-e422-42e0-a05a-86144428c5b9.png)

Binding errors
```
Error: BindingExpression path error: 'UserName' property not found on 'Uno.Extensions.Reactive.Core.StateImpl`1[Ligric.UI.Infrastructure.Settings.Credentials]'. BindingExpression: Path='Credentials.UserName' DataItem='Ligric.UI.Infrastructure.Presentation.AuthorizationViewModel'; target element is 'Microsoft.UI.Xaml.Controls.TextBox' (Name='null'); target property is 'Text' (type 'String')

Error: BindingExpression path error: 'Password' property not found on 'Uno.Extensions.Reactive.Core.StateImpl`1[Ligric.UI.Infrastructure.Settings.Credentials]'. BindingExpression: Path='Credentials.Password' DataItem='Ligric.UI.Infrastructure.Presentation.AuthorizationViewModel'; target element is 'Microsoft.UI.Xaml.Controls.PasswordBox' (Name='null'); target property is 'Password' (type 'String')

```

**Comment by jeromelaban** [comment]
@dr1rrb will be able to help

**Reply by limeniye** [reply]
@dr1rrb , I'm glad to get your help ^^

**Reply by francoistanguay** [reply]
How do you instantiate the ViewModel andset it in the View's DataContext? There's a generated BindableXYZModel that is code generated and that's what needs to be set in the View.

Are you using Uno Navigation as well or just Reactive?

**Reply by francoistanguay** [reply]
Docs aren't super clear but some info here: https://platform.uno/docs/articles/external/uno.extensions/doc/Overview/Reactive/in-apps.html

**Reply by limeniye** [reply]
@francoistanguay , thank you for your answer.

```
"How do you instantiate the ViewModel andset it in the View's DataContext?"
```

I registred it inside my [host file](https://github.com/ligric/ligricui/blob/main/src/Ligric.UI.Uno.Shared/App.xaml.host.cs#L80-L83):
```C#
        views.Register(
            new ViewMap<AuthorizationPage, AuthorizationViewModel>(),
            new ViewMap(ViewModel: typeof(ShellViewModel))
        );
```

This is my TextBox inside [xaml file:](https://github.com/ligric/ligricui/blob/main/src/Ligric.UI.Uno.Shared/Views/AuthorizationPage.xaml#L38)
```XML
    <TextBox Text="{Binding Credentials.UserName, Mode=TwoWay, UpdateSourceTrigger=PropertyChanged}"/>
```

This is Credentials property inside my [ViewModel](https://github.com/ligric/ligricui/blob/main/src/Ligric.UI.Infrastructure/Presentation/AuthorizationViewModel.cs#L20):
```C#
    public IState<Credentials> Credentials => State<Credentials>.Empty(this);
```

**Reply by nickrandolph** [reply]
Change this line 
https://github.com/ligric/ligricui/blob/ba6b895d04acd8dbe797aa3853427b8c25abd40b/src/Ligric.UI.Uno.Shared/App.xaml.host.cs#L67
to:
`.UseNavigation(Ligric.UI.Infrastructure.ReactiveViewModelMappings.ViewModelMappings, RegisterRoutes)`

**Reply by limeniye** [reply]
@nickrandolph, Thank you.
Yes. You were right.
This is an answer.
